keps/prod-readiness/sig-network/2086.yaml

```diff
@@ -0,0 +1,3 @@
+kep-number: 2086
+alpha:
+  approver: "@johnbelamaric"
```

Comment:
I actually claim we should rethink this proposal with the new approach that we're proceed with decribed in:
https://github.com/kubernetes/enhancements/pull/2434

The arguments described in the proposal (EndpointSlice per node) will no longer hold with that (we can easily imagine "forNode" hint too).
So instead of rushing with this one, I would really like to take a step back and ensure that we can't converge this with #2434 

@thockin @robscott @aojea @johnbelamaric

Reply:
the slices have already a `hostname` field, IIUIC this is a just a matter of signalling kube-proxy to leverage it, it can be an annotation or an api field, no?

```
	// hostname of this endpoint. This field may be used by consumers of
	// endpoints to distinguish endpoints from each other (e.g. in DNS names).
	// Multiple endpoints which use the same hostname should be considered
	// fungible (e.g. multiple A values in DNS). Must pass DNS Label (RFC 1123)
	// validation.
	// +optional
	Hostname *string
```

Reply:
So, the problem is that the API field can be contradictory with the new annotations `topology.kubernetes.io/topology-aware-routing`, right?

Maybe we should add a new value to the annotation to implement local traffic ...

Reply:
Good point that we don't need any additional hints for that.

Regarding API - given it's explicitly stated there as "local-traffic will require careful handing" (or sth like that), I think we should think if it doesn't make sense to unify (using single annotation is what i was thinking, but I think we should rely on API approvers (@thockin - back to you) to recommend if that makes sense).

Reply:
> Regarding API - given it's explicitly stated there as "local-traffic will require careful handing" (or sth like that), I think we should think if it doesn't make sense to unify (using single annotation is what i was thinking, but I think we should rely on API approvers (@thockin - back to you) to recommend if that makes sense).

There was a lot of back and forth on this and the conclusion IIRC is that "route to node-local" is a distinct and common enough use-case to warrant a different API from topology-aware routing. More on this email thread: https://groups.google.com/g/kubernetes-sig-network/c/wXd1D_fKjqU. But that was also under the assumption that we were going to use EndpointSlice subsetting and a EndpointSlice per-node didn't make sense. 

> So, the problem is that the API field can be contradictory with the new annotations topology.kubernetes.io/topology-aware-routing, right?

There is an overlap if you consider `Local` or `PreferLocal`. The assumption so far has been that this annotation would only be used with `Cluster` (now `All`). My personal preference would actually be to have `TopologyAware` be another possible value of `internalTrafficPolicy` as opposed to an annotation but I'll leave that feedback in the other KEP PR.

Reply:
OK, Let's untangle.

We have `nodeName` as a per-endpoint field (gated by EndpointSliceNodeName, alpha in 1.20).  @robscott should we move that beta in 21?

Proposal #2367 adds a `zone` field and EOLs `topology`.

With those two, we know where each endpoint is in the topology.

I want topology, in general, to be automatic (eventually) in most cases.  That primarily means zonal affinity, but it's not beyond reason that we could automatically engage nodal affinity if we know every node has an endpoint.  That means that, normally, we want no API.

We know we need SOME form of API a) until we feel comfortable changing the default from opt-in to opt-out; and b) to allow opt-in/out.  Will we EVER feel comfortable changing it?  I don't know.  We don't generally change API defaults except at major API versions.  I want to see the default be the happy path, but given the potential for breakage, I am not sure.

So now we have 2 APIs - one to enable topology hints (#2434 `annotations[service.kubernetes.io/topology-aware-routing]`) and this (`internalTrafficPolicy`).  They do really seem to overlap, and I can't imagine when they would both be useful at the same time.  Great flag, @wojtek-t - thanks.

@robscott @swetharepakula - What do you think about the idea?  We could extend the values here from (Cluster, Local, PreferLocal) to include "Topology".  Or maybe we should think about all the names, to make sure they best capture the expanded semantics unambiguously.  E.g. Cluster, SameNode, PreferSameNode, TopologyOptimized.  Open question - do we want "Auto" to mean "decide for me"?

I think we can get these KEPs approved without merging them just yet, but good to discuss now.  If we do it, we probably need to so it for 21.  Is there major risk in linking them? Worst case if one fails?

Reply:
I think the easiest thing to understand here might be adding `PreferZone` and potentially `RequireZone` as options for `InternalTrafficPolicy`. A couple potential downsides:

1. EndpointSlice hints are only generated when `internalTrafficPolicy == PreferZone`. Not the end of the world, but feels a little strange that there's not just one thing to say whether or not hints should exist.
2. Limiting this to `internalTrafficPolicy` removes a potential link to LB config. If there were a separate annotation/field specific to topology it might be able to apply more broadly.

I don't think either of those are necessarily blockers though. Maybe a slightly different field name would help to reflect the slightly broader purpose here? There is a lot of overlap here so it would be good to merge config together if possible.

Reply:
>We have nodeName as a per-endpoint field (gated by EndpointSliceNodeName, alpha in 1.20). @robscott should we move that beta in 21?

Yes, absolutely. That's already covered in the plans for 1.21 in the EndpointSlice KEP.

Reply:
cc @maplain who is working on the initial implementation for internalTrafficPolicy https://github.com/kubernetes/kubernetes/pull/96600

Reply:
I think it makes sense to try combine these. There is a lot of overlap. I prefer `PreferZone` over `Topology` as I

In regards to "Auto" meaning "decide for me"- IIUC that means that based on the endpoints available, choosing between `Cluster`, `Local`, `PreferLocal`, and `PreferZone`. Would that require an additional field or annotation to state what was chosen? `PreferZone` is the only one that would be detectable because of the the `topologyHints`. How would kube-proxy differentiate between the others?

Reply:
> I think the easiest thing to understand here might be adding PreferZone 

The semantics of PreferNode are "if there is an endpoint on-node, use it.  Only
if there is no endpoint should you overflow to another node".  I don't think we
want the same for zone (or at least, that's not the heuristic you have
developed and almost certainly not the best for "auto").

> potentially RequireZone as options for InternalTrafficPolicy

If and when we have use-cases.

> EndpointSlice hints are only generated when internalTrafficPolicy == PreferZone

We could always generate hints, but why bother unless we are in a mode where
they matter?  We might want external traffic to have topology, too?

> Limiting this to internalTrafficPolicy removes a potential link to LB config.

Actually - is that a reason why topology and local-only might not be the same?
What happens if I set `internalTrafficPolicy: Topology` and I take external
traffic?  Does that also follow topology?  Or does topological subsetting apply
to all "whole cluster" policies?

In the context of external traffic, "Local" has a semantic implication (no
SNAT) as well as a perfomance/cost implication.  We have to respect that.  But
the default is "Cluster" and it seems to me that topology should apply in those
cases (cost/perf win but no semantic change).

This point is weighing on me - thoughts?

> In regards to "Auto" meaning "decide for me"- IIUC that means that based on the endpoints available, choosing between Cluster, Local, PreferLocal, and PreferZone. Would that require an additional field or annotation to state what was chosen? PreferZone is the only one that would be detectable because of the the topologyHints. How would kube-proxy differentiate between the others?

I'm not sure it's the right approach, but "Auto" could look at the endpoints
and decide "I have at least one endpoint on every node, so I will set a hint to
use the same node" vs "I don't have enough for node, but I can bound by zone" or
even some mixed mode.

Reply:
Updates from Slack thread:

- rename `internalTrafficPolicy` to `trafficPolicy` to reflect increased scope
- if `externalTrafficPolicy=Cluster`, fall back to `trafficPolicy` for external sources
- if `externalTrafficPolicy=Local` -- this rule only takes precedent for external sources, with internal traffic following `trafficPolicy`